apps/websocket/internal/index.js

```diff
@@ -0,0 +1,47 @@
+import validTopic from './valid-topic';
+import createSubscriptions from './subscriptions';
+
+const defaultURL = `ws://${location.hostname}:8000`;
+
+const logger = debug => (...args) => {
+  if (debug) {
+    console.log(...args);
+  }
+};
+
+const createWebsocket = ({ url = defaultURL, debug = false }) => {
+  const ws = new WebSocket(url);
+  const log = logger(debug);
+
+  const subscriptions = createSubscriptions(log);
+
+  const onOpen = (event) => log('connected', event);
+
+  const onMessage = (event) => {
+    const { topic, payload } = JSON.parse(event.data);
+
+    subscriptions.matching(topic).forEach(cb => cb({ topic, payload }));
+  };
+
+  const subscribe = (topic, cb) => {
+    subscriptions.subscribe(topic, cb);
```

Comment:
Does this imply clients can only have one callback per topic? If so, if a client calls subscribe again, is it obvious that their last callback will no longer work?

Reply:
I could either:

- Allow multiple callbacks
- Throw an error when trying to set an existing sub

Which would you prefer? Multiple callbacks might be a bit confusing?

Reply:
I guess warning for now would make sense. I just feel like it could be the source of some confusing bugs.